randovania/generator/filler/hint_placement.py

```diff
@@ -0,0 +1,198 @@
+    PickupIndex(115): HintType.GUARDIAN, # Annihilator Beam (Quadraxis)
+}
+
+T = TypeVar("T")
+
+
+def _create_weighted_list(rng: Random,
+                          current: List[T],
+                          factory: Callable[[], List[T]],
+                          ) -> List[T]:
+    """
+    Ensures we always have a non-empty list.
+    :param rng:
+    :param current:
+    :param factory:
+    :return:
+    """
+    if not current:
+        current = factory()
+        rng.shuffle(current)
+
+    return current
+
+
+def _hint_for_index(index: PickupIndex) -> Hint:
+    if index in _SPECIAL_HINTS:
+        return Hint(_SPECIAL_HINTS[index], PrecisionPair.detailed(), index)
+    else:
+        return Hint(HintType.LOCATION, None, index)
+
+
+def add_hint_precisions(patches: GamePatches, rng: Random) -> GamePatches:
+    """
+    Adds precision to all hints that are missing one.
+    :param patches:
+    :param rng:
+    :return:
+    """
+
+    hints_to_replace = {
+        asset: hint
+        for asset, hint in patches.hints.items()
+        if hint.precision is None
+    }
+
+    asset_ids = list(hints_to_replace.keys())
+
+    # Add random precisions
+    rng.shuffle(asset_ids)
+    precisions = []
+    for asset_id in asset_ids:
+        precisions = _create_weighted_list(rng, precisions, PrecisionPair.weighted_list)
+        precision = precisions.pop()
+
+        hints_to_replace[asset_id] = dataclasses.replace(hints_to_replace[asset_id], precision=precision)
+
+    # Replace the hints the in the patches
+    return dataclasses.replace(patches, hints={
+        asset: hints_to_replace.get(asset, hint)
+        for asset, hint in patches.hints.items()
+    })
+
+
+def replace_hints_without_precision_with_jokes(patches: GamePatches) -> GamePatches:
+    """
+    Adds WRONG_GAME precision to all hints that are missing one precision.
+    :param patches:
+    :return:
+    """
+
+    hints_to_replace = {
+        asset: dataclasses.replace(hint, precision=PrecisionPair.joke())
+        for asset, hint in patches.hints.items()
+        if hint.precision
+    }
+
+    return dataclasses.replace(patches, hints={
+        asset: hints_to_replace.get(asset, hint)
+        for asset, hint in patches.hints.items()
+    })
+
+
+def place_hints(configuration: LayoutConfiguration, final_state: State, patches: GamePatches, rng: Random,
+                world_list: WorldList, status_update: Callable[[str], None]) -> GamePatches:
+    status_update("Placing hints...")
+
+    state_sequence = [final_state]
+    while state_sequence[-1].previous_state:
+        state_sequence.append(state_sequence[-1].previous_state)
+    state_sequence = tuple(reversed(state_sequence))
+
+    possible_hint_indices = {index for index, pickup in patches.pickup_assignment.items()
+                             if pickup.can_get_hint or index in _SPECIAL_HINTS}
+
+    sequence: List[Union[PickupIndex, LogbookAsset]] = []
+    for state in state_sequence[1:]:
```

Comment:
This for loop could be refactored out into another function.

Reply:
✓